crates/btc-relay/src/lib.rs

```diff
@@ -1218,6 +1271,115 @@ impl<T: Config> Pallet<T> {
         Ok(())
     }
 
+    /// The target value represents the mining difficulty required to mine a block with the specified hash.
+    ///
+    /// # Parameters
+    ///
+    /// * `block_hash`: A 256-bit block hash.
+    ///
+    /// # Returns
+    ///
+    /// * `Result<u32, DispatchError>`: A `Result` containing the retrieved target value as a `u32` if successful, or a
+    ///   `DispatchError` if an error occurs during the retrieval.
+    fn get_target(block_hash: H256Le) -> Result<u32, DispatchError> {
+        let block_header = Self::get_block_header_from_hash(block_hash)?;
+        // Fixme: Should this be kept as a U256 type and other u32 type can be converted to U256?
+        // Currently taking low_u32/as_u32 loses precision
+        Ok(block_header.block_header.target.low_u32())
+    }
+
+    /// Calculates the chainwork of a block header at a given height based on the provided block hash.
+    /// Reference: https://github.com/spesmilo/electrum/blob/cee22abcb5544c5a6fa7f8a8108ccda9c32c2e29/electrum/blockchain.py#L582-L588
+    ///
+    /// # Arguments
+    ///
+    /// * `block_hash`: A 256-bit block hash.
+    ///
+    /// # Returns
+    ///
+    /// * `Result<u32, DispatchError>`: A `Result` containing the calculated chainwork as a `u32` if successful, or a
+    ///   `DispatchError` if an error occurs during target retrieval.
+    fn chainwork_of_header_at_height(block_hash: H256Le) -> Result<u32, DispatchError> {
+        let target = Self::get_target(block_hash)?;
+        let work = (2_u32.saturating_pow(256).saturating_sub(target).saturating_less_one())
+            .saturating_div(target.saturating_plus_one())
+            .saturating_plus_one();
+        Ok(work)
+    }
+
+    /// The chainwork is a measure of the total computational work done by the chain up to a certain point.
+    /// Reference: https://github.com/spesmilo/electrum/blob/cee22abcb5544c5a6fa7f8a8108ccda9c32c2e29/electrum/blockchain.py#L589-L614
+    ///
+    /// # Parameters
+    ///
+    /// * `chain`: new blockchain element
+    ///
+    /// # Returns
+    ///
+    /// * `Result<u32, DispatchError>`: A `Result` containing the calculated cumulative chainwork as a `u32` if
+    ///   successful, or a `DispatchError` if an error occurs during the calculation.
+    fn get_chainwork(chain: BlockChain) -> Result<u32, DispatchError> {
```

Comment:
I'm not sure it's a good idea to use the electrum function unchanged. In fact, I don't think it'll work, because
- (1) it's super expensive - it iterates all the way from block zero. We might exceed the block time because of all the iteration
- (2) We didn't initialize the relay at height 0. The `chainwork_of_header_at_height` function would fail.

This is a pretty high-risk change, and I was ask you to test it thoroughly. If I'm right and this doesn't work, it could've been caught with a chopsticks test.

Reply:
yes, reviewing the code back again even if the loop over the target it would still take `btc_best_block/2016` ie `808,849
/2016 = 402` iterations. I will still do the chopsticks test and put the results in the comment.